`ConfigParser.optionxform` assignment flagged as “shadowing” (false positive)

### Summary

```python
from configparser import ConfigParser

cfgparser = ConfigParser()
cfgparser.optionxform = str  # <--- Should be allowed, per Python docs
``` 
Ref. https://docs.python.org/3/library/configparser.html#configparser.ConfigParser.optionxform

### ty

```
Implicit shadowing of function `optionxform` (invalid-assignment) [Ln 4, Col 1]
``` 

### pyrefly


```
ERROR sandbox.py:4:25-28: `type[str]` is not assignable to attribute `optionxform` with type `BoundMethod[ConfigParser, (self: ConfigParser, optionstr: str) -> str]` [[bad-assignment](https://pyrefly.org/en/docs/error-kinds/#bad-assignment)]
  Positional parameter name mismatch: got `object`, want `optionstr`
``` 

### pyright

```
Cannot assign to attribute "optionxform" for class "ConfigParser"
  No overloaded function matches type "(optionstr: str) -> str"  (reportAttributeAccessIssue)
``` 

### mypy

```
main.py:4: error: Cannot assign to a method  [method-assign]
main.py:4: error: Incompatible types in assignment (expression has type "type[str]", variable has type "Callable[[str], str]")  [assignment]
Found 2 errors in 1 file (checked 1 source file)
``` 

### Version

0.0.2

--- a/crates/ty_python_semantic/resources/mdtest/diagnostics/shadowing.md
+++ b/crates/ty_python_semantic/resources/mdtest/diagnostics/shadowing.md
@@ -42,3 +42,21 @@ error[invalid-assignment]: Object of type `Literal[1]` is not assignable to `def
   |
 info: Implicit shadowing of function `f`. Add an annotation to make it explicit if this is intentional
 ```
+
+## No implicit shadowing for attribute assignments
+
+```py
+from configparser import ConfigParser
+
+config = ConfigParser()
+config.optionxform = str  # snapshot: invalid-assignment
+```
+
+```snapshot
+error[invalid-assignment]: Object of type `<class 'str'>` is not assignable to attribute `optionxform` of type `def optionxform(self, optionstr: str) -> str`
+ --> src/mdtest_snippet.py:4:1
+  |
+4 | config.optionxform = str  # snapshot: invalid-assignment
+  | ^^^^^^^^^^^^^^^^^^
+  |
+```

--- a/crates/ty_python_semantic/src/types/diagnostic.rs
+++ b/crates/ty_python_semantic/src/types/diagnostic.rs
@@ -1423,29 +1423,10 @@ pub(crate) fn is_invalid_typed_dict_literal(
 fn report_invalid_assignment_with_message<'db, 'ctx: 'db, T: Ranged>(
     context: &'ctx InferContext,
     node: T,
-    target_ty: Type<'db>,
     message: std::fmt::Arguments,
 ) -> Option<LintDiagnosticGuard<'db, 'ctx>> {
     let builder = context.report_lint(&INVALID_ASSIGNMENT, node)?;
-
-    let mut diag = builder.into_diagnostic(message);
-
-    match target_ty {
-        Type::ClassLiteral(class) => {
-            diag.info(format_args!(
-                "Implicit shadowing of class `{}`. Add an annotation to make it explicit if this is intentional",
-                class.name(context.db()),
-            ));
-        }
-        Type::FunctionLiteral(function) => {
-            diag.info(format_args!(
-                "Implicit shadowing of function `{}`. Add an annotation to make it explicit if this is intentional",
-                function.name(context.db()),
-            ));
-        }
-        _ => {}
-    }
-    Some(diag)
+    Some(builder.into_diagnostic(message))
 }
 
 pub(super) fn note_numbers_module_not_supported<'db>(
@@ -1619,7 +1600,6 @@ pub(super) fn report_invalid_assignment<'db>(
     let Some(mut diag) = report_invalid_assignment_with_message(
         context,
         diagnostic_range,
-        target_ty,
         format_args!(
             "Object of type `{}` is not assignable to `{}`",
             value_ty.display_with(context.db(), settings.clone()),
@@ -1628,6 +1608,24 @@ pub(super) fn report_invalid_assignment<'db>(
     ) else {
         return;
     };
+
+    if matches!(target_node, AnyNodeRef::ExprName(_)) {
+        match target_ty {
+            Type::ClassLiteral(class) => {
+                diag.info(format_args!(
+                    "Implicit shadowing of class `{}`. Add an annotation to make it explicit if this is intentional",
+                    class.name(context.db()),
+                ));
+            }
+            Type::FunctionLiteral(function) => {
+                diag.info(format_args!(
+                    "Implicit shadowing of function `{}`. Add an annotation to make it explicit if this is intentional",
+                    function.name(context.db()),
+                ));
+            }
+            _ => {}
+        }
+    }
 
     if value_node.is_some() {
         match definition_kind {
@@ -1681,7 +1679,6 @@ pub(super) fn report_invalid_attribute_assignment(
     let Some(mut diag) = report_invalid_assignment_with_message(
         context,
         range,
-        target_ty,
         format_args!(
             "Object of type `{}` is not assignable to attribute `{attribute_name}` of type `{}`",
             source_ty.display(context.db()),

